main.py

```diff
@@ -1511,43 +1512,90 @@ def show_hover(self, point, contents):
             max_width=800)
 
 
-class CompletionHandler(sublime_plugin.EventListener):
-    def __init__(self):
-        self.completions = []  # type: List[Tuple[str, str]]
-        self.refreshing = False
+class CompletionState(object):
+    IDLE = 0
+    REQUESTING = 1
+    APPLYING = 2
+    CANCELLING = 3
 
-    def on_query_completions(self, view, prefix, locations):
-        if not is_supported_view(view):
-            return None
 
-        if not self.refreshing:
-            client = client_for_view(view)
+class CompletionHandler2(sublime_plugin.ViewEventListener):
+    def __init__(self, view):
+        self.view = view
+        self.autocomplete_triggers = None
+        self.completions = []  # type: List[Tuple[str, str]]
+        self.state = CompletionState.IDLE
+        self.next_request = None
 
-            if not client:
-                return
+    @classmethod
+    def is_applicable(cls, settings):
+        syntax = settings.get('syntax')
+        return is_supported_syntax(syntax)
 
-            completionProvider = client.get_capability('completionProvider')
-            if not completionProvider:
+    def initialize_triggers(self):
+        client = client_for_view(self.view)
+        if client:
+            completionProvider = client.get_capability(
+                'completionProvider')
+            if completionProvider:
+                self.autocomplete_triggers = completionProvider.get(
+                    'triggerCharacters')
                 return
 
-            autocomplete_triggers = completionProvider.get('triggerCharacters')
+        self.autocomplete_triggers = []
 
-            if locations[0] > 0:
-                self.completions = []
-                prev_char = view.substr(
-                    sublime.Region(locations[0] - 1, locations[0]))
-                if prev_char not in autocomplete_triggers:
-                    return None
+    def is_after_trigger_character(self, location):
+        if location > 0:
+            prev_char = self.view.substr(
+                sublime.Region(location - 1, location))
```

Comment:
You can use `sublime.Region(location - 1)` I believe.

Reply:
You can even call `self.view.substr(location - 1)` as `substr` understands both a `Region` and a `Point`